### PR TITLE
feat: introduce `MismatchedTagPatternArity`

### LINKED ISSUE #9381: Improve pattern matching arity error
```
enum E[one, two] { case A(one, two) }

def example(): Unit = match ??? {
    case E.A(one, two, three) => ()
}
```
```
>> Unable to unify the types: '? -> (? -> ? \ ?)' and '? -> (? -> (? -> ?))'.

4 |     case E.A(one, two, three) => ()
             ^^^^^^^^^^^^^^^^^^^^
             mismatched types.

Type One: t0 -> (t1 -> E[t0, t1])
Type Two: t0 -> (t1 -> (t2 -> t3))
```

### PR DESCRIPTION
**Pull request by JonathanStarup**
Fixes #9381

```
def foo(l: List[Int32]): Int32 = match l {
  case Nil => 42
  case Cons(x, _xs, _bonus) => x
}
```
This has the following error on master:
```
-- Type Error -------------------------------------------------- test.flix

>> Unable to unify the types: '? -> (? -> ? \ ?)' and '? -> (? -> (? -> ?))'.

3 |   case Cons(x, _xs, _bonus) => x
           ^^^^^^^^^^^^^^^^^^^^
           mismatched types.

Type One: t0 -> (List[t0] -> List[t0])
Type Two: Int32 -> (t0 -> (t1 -> List[Int32]))
```

But with this PR it is (notice that type checking succeeds since I correct the pattern arity)
```
-- Resolution Error -------------------------------------------------- test.flix

 Expected 2 arguments for List.Cons but found 3 arguments.

3 |   case Cons(x, _xs, _bonus) => x
           ^^^^^^^^^^^^^^^^^^^^
           mismatched arguments
```

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala [hunk 1]
```diff
@@ -1842,7 +1842,17 @@ object Resolver {
...
           case Result.Ok(c) =>
-            val ps = pats.map(visit)
+            val ps0 = pats.map(visit)
+            // If there is not a pattern for each term, give an error.
+            val ps = if (ps0.lengthCompare(c.tpes) != 0) {
+              val expectedTerms = c.tpes.length
+              val error = ResolutionError.MismatchedTagPatternArity(c.sym, expected = expectedTerms, actual = ps0.length, loc)
+              sctx.errors.add(error)
+              // maintain as many sensible arguments to allow further type checking, etc.
+              ps0.take(expectedTerms).padTo(expectedTerms, ResolvedAst.Pattern.Wild(loc.asSynthetic))
```
Comment (JonathanStarup) [inline main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:1852]: If this padding/trimming is not done, then the user will see both the new pretty error and the old type error. With this added you might now just get undefined variables (gave too many, so the last extra are ignored)

File: main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala [hunk 2]
```diff
@@ -534,6 +534,37 @@ object ResolutionError {
     override def explain(formatter: Formatter): Option[String] = None
   }
 
+  /**
+    * An error indicating the number of term patterns does not match the enum tag term count.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala:538]: -> An error indicating that the number of arguments to a tag-pattern is mismatched.

File: main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala [hunk 3]
```diff
@@ -534,6 +534,37 @@ object ResolutionError {
...
+    *
+    * @param caze     the enum case symbol.
+    * @param expected the expected number of terms.
+    * @param actual   the actual number of patterns.
+    * @param loc      the location where the error occurred.
+    */
+  case class MismatchedTagPatternArity(caze: Symbol.CaseSym, expected: Int, actual: Int, loc: SourceLocation) extends ResolutionError {
+    override def summary: String = s"Expected ${Grammar.n_things(expected, "term")} but found $actual."
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala:546]: term -> arguments

File: main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala [hunk 4]
```diff
@@ -534,6 +534,37 @@ object ResolutionError {
...
+
+    /**
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala:548]: Drop doc comment

File: main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala (hunk 3 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala:546]: ... but found $ actual ${Grammar.n_things(expected, "argument")}

File: main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala [hunk 5]
```diff
@@ -534,6 +534,37 @@ object ResolutionError {
...
+      * Returns the formatted error message.
+      */
+    override def message(formatter: Formatter): String = messageWithLink {
+      import formatter.*
+      s""">> Mismatched arity.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala:553]: -> Expected ${Grammar.n_things(expected, "argument")} for Cyan(caze) but found ${${Grammar.n_things(actual, "argument")}} arguments.

File: main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala [hunk 6]
```diff
@@ -534,6 +534,37 @@ object ResolutionError {
...
+         |
+         |The enum case $caze expects ${Grammar.n_things(expected, "term")},
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala:555]: Drop this

File: main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala [hunk 7]
```diff
@@ -534,6 +534,37 @@ object ResolutionError {
...
+         |but ${Grammar.n_are(actual)} provided here.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala:556]: Drop

File: main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala [hunk 8]
```diff
@@ -534,6 +534,37 @@ object ResolutionError {
...
+         |
+         |${code(loc, s"expected ${Grammar.n_things(expected, "term")} but found $actual")}
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala:558]: -> mismatched arguments

File: main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala [hunk 9]
```diff
@@ -534,6 +534,37 @@ object ResolutionError {
...
+         |""".stripMargin
+    }
+
+    /**
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala:562]: Dont override

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala [hunk 10]
```diff
@@ -1842,7 +1842,17 @@ object Resolver {
       case NamedAst.Pattern.Tag(qname, pats, loc) =>
         lookupTag(qname, env, ns0, root) match {
           case Result.Ok(c) =>
-            val ps = pats.map(visit)
+            val ps0 = pats.map(visit)
+            // If there is not a pattern for each term, give an error.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:1846]: // Check if the number of arguments match

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:1852]: Why Pattern.Wild and not Pattern.Error?

File: main/test/ca/uwaterloo/flix/language/phase/TestResolver.scala [hunk 11]
```diff
@@ -1627,6 +1627,44 @@ class TestResolver extends AnyFunSuite with TestUtils {
     expectError[ResolutionError.MismatchedOpArity](result)
   }
 
+  test("Test.WrongTagPattenCount.01") {
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/phase/TestResolver.scala:1630]: test name should be the samae as the error

File: main/test/ca/uwaterloo/flix/language/phase/TestResolver.scala [hunk 12]
```diff
@@ -1627,6 +1627,44 @@ class TestResolver extends AnyFunSuite with TestUtils {
...
+        |      case Nil => 42
+        |      case Cons(_) => 42
+        |    }
+        |}
+        |""".stripMargin
+    val result = compile(input, Options.TestWithLibNix)
+    expectError[ResolutionError.MismatchedTagPatternArity](result)
+  }
+
+  test("Test.WrongTagPattenCount.02") {
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/phase/TestResolver.scala:1649]: test name should be the samae as the error

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala (hunk 1 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:1852]: fixed

**Comment by JonathanStarup** [comment]
@magnus-madsen fixed

**Review by magnus-madsen — APPROVED** [review]

**Pull request merged**